### ISSUE TITLE
Comment fonctionne la détection de résolveur ?

### ISSUE THREAD

**Issue by bortzmeyer** [changetondns/changetondns.fr#9]
Mon Firefox utilise le résolveur DoH de FDN et pourtant ChangeTonDNS affiche "Serveur DNS

OVH SAS

Non Vulnérable"

OVH est bien le fournisseur d'accès mais je m'attendais à voir de l'information sur mon résolveur, pas sur mon FAI.

**Comment by bortzmeyer** [comment]
Je soupçonne qu'il y a en fait deux problèmes:

- le code de détection utilise l'adresse IP du client HTTP, pas celle du résolveur,
- le code de détection ne teste pas réellement le comportement du résolveur (censure ou pas), juste le préfixe IP et ne fait donc pas la différence, par exemple, entre les résolveurs officiels d'Orange et un résolveur tiers, non menteur, hébergé chez Orange.

**Comment by matrixbx** [comment]
Bonjour Stéphane,

de ce que je constate, un http get sur "1.1111111.bash.ws" pour déclencher un lookup dns par le résolveur utilisé (qui répond tjrs NXDOMAIN).
suivi d'un http get sur "https://bash.ws/dnsleak/test/1111111?json".

Le json renvoyé contient d'abord l'ip publique visible du client,
puis l'ip publique visible du résolveur ayant interrogé le sous domaine en question (qui varie à chaque test).

```
host 1.1234567.bash.ws 45.67.81.23
curl -s 'https://bash.ws/dnsleak/test/1234567?json' |jq '.'
  {
    "ip": "45.67.81.21",
    "country": "fr",
    "country_name": "France, French Republic",
    "asn": "AS198985 Aquilenet",
    "type": "dns"
  },
```

```
host 1.2345678.bash.ws 80.67.169.12
  {
    "ip": "80.67.169.16",
    "country": "fr",
    "country_name": "France, French Republic",
    "asn": "AS20766 Association Gitoyen",
    "type": "dns"
  },
```

**Comment by bortzmeyer** [comment]
Oui,mais ce n'est pas ce qui est affiché à l'utilisateur.

**Comment by Jourdelune** [comment]
Une fois la requête faite à l'api on regarde simplement si le nom du FAI français est présent, cela peut être améliorer.

**Comment by Jourdelune** [comment]
Une fois la requête faite à l'api on regarde simplement si le nom d'un FAI français est présent, cela peut être amélioré.

**Comment by matrixbx** [comment]
Pas évident d'être plus précis, mais on peut essayer.
La liste des résolveurs des FAIs mis a disposition de leurs clients n'est pas bien difficile à établir, le souci est que le résolveur interrogé par le client, n'est pas forcément celui qui interrogera au final (et que ça peut changer sans trop prévenir).

En réalisant quelques tests, j'ai établi que pour Free, ceux qui interrogent sont :
213.228.60.{129..144} et 2a01:e00:ffff:53::{1..16}

Et pour Orange/Oleane/OBS : 
80.10.201.{16..126}, 80.10.203.{16..126}, 90.84.{8..9}.{1..111}, 195.25.5.{31..45}, et 213.56.27.{71..85}

À compléter avec ceux de  Bouygues et SFR

**Comment by Jourdelune** [comment]
Une autre approche serait de faire une requête à des noms de domaines bloqués en France et selon la réponse déterminer si l'utilisateur est vulnérable ou non, même si l'identification du DNS utilisé par le client peut être erronés (ce n'es pas si grave, le plus important est de savoir si l'utilisateur est vulnérable à un dns menteur). Votre approche est également intéressante.

**Comment by matrixbx** [comment]
Remarque sur l'oiseau bleu : 
ne pas alerter trop fort quand l'ip du client et du résolveur est la même, c'est quand même le signe assez limpide d'un résolveur local.

**Comment by Jourdelune** [comment]
Oui déjà on va enlever le signalement quand l'utilisateur est également le resolveur.

**Comment by ostix360** [comment]
> de ce que je constate, un http get sur "1.1111111.bash.ws" pour déclencher un lookup dns par le résolveur utilisé (qui répond tjrs NXDOMAIN). suivi d'un http get sur "https://bash.ws/dnsleak/test/1111111?json".
> 
> Le json renvoyé contient d'abord l'ip publique visible du client, puis l'ip publique visible du résolveur ayant interrogé le sous domaine en question (qui varie à chaque test).
> 
> ```
> host 1.1234567.bash.ws 45.67.81.23
> curl -s 'https://bash.ws/dnsleak/test/1234567?json' |jq '.'
>   {
>     "ip": "45.67.81.21",
>     "country": "fr",
>     "country_name": "France, French Republic",
>     "asn": "AS198985 Aquilenet",
>     "type": "dns"
>   },
> ```
> 
> ```
> host 1.2345678.bash.ws 80.67.169.12
>   {
>     "ip": "80.67.169.16",
>     "country": "fr",
>     "country_name": "France, French Republic",
>     "asn": "AS20766 Association Gitoyen",
>     "type": "dns"
>   },
> ```

Effectivement pour l'instant nous vérifions si dans le nom du asn il y a le nom de ces fai `['bouygues', 'free sas', 'sfr', 'orange']` 
Forcément tout asn ne faisant pas parti de cette liste sont considéré comme non vulnérable

**Comment by ostix360** [comment]
> Remarque sur l'oiseau bleu : ne pas alerter trop fort quand l'ip du client et du résolveur est la même, c'est quand même le signe assez limpide d'un résolveur local.

https://github.com/changetondns/changetondns.fr/commit/ba5f4a1af78c0045e68dbdf3bf02394ccc96d904

Je vous invite à essayer si c'est bien le cas sur https://dev.changetondns-fr.pages.dev/

**Comment by matrixbx** [comment]
Le 1er test a échoué, l'interrogation client se faisant en IPv6 et la résolution en IPv4.
Le second a réussi, car en IPv6 dans les 2 cas.

**Comment by ostix360** [comment]
Vous pouvez réessayer.
Malheureusement je ne peux pas essayer pas moi même, je ne peux pas utiliser l'IPv6 sur ma machine.
Merci à vous

**Comment by matrixbx** [comment]
Remarqué  un comportement surprenant, parfois ça s'affole et ça lance des centaines si ce n'est des milliers de requêtes et le navigateur (brave) mouline très fort...

Sinon chez Orange.
En IPv4 ou v6, si l'IP du client et du résolveur est la même ça indique bien "Non vulnérable".
Reste que si le résolveur est sur le réseau local mais pas sur la machine client, il n'a pas la même IPv6 publique et le test échoue, il faudrait limiter la comparaison en IPv6 à la taille du préfixe.

Reste les cas où le client est en IPv4 et le résolveur en IPv6 (ou inversement) qui échouent très logiquement, il faudra probablement utiliser la méthode des adresses de résolveurs connu pour êtres menteurs et simplement émettre un avertissement léger en mode "doute" sinon.

**Comment by Jourdelune** [comment]
D'accord mais pouvez vous compléter la liste pour Bouygues et SFR ? Ou sinon expliquez comment vous avez étez en mesure de déterminer quel ip est utilisé par ces résolveurs ? De cette manière :
- Si l'ip du résolveur est contenus dans ce filtre, afficher un warning
- Si l'ip du résolveur n'est pas contenus dans ce filtre mais que l'ip appartient à un FAI, mettre un warning orange
- Sinon rien.

**Comment by matrixbx** [comment]
Non, je ne peux pas, ne disposant pas d'accès chez ces FAIs, qui comme les autres, ne répondent qu'à leurs IPs.

Par exemple pour Bouygues (et sous réserve que les IP des résolveurs soient les bonnes, à adapter sinon).
```
LOGFILE="Resolveurs_Bouygues" # le nom qui vous sied.

Bouygues=(194.158.122.1{0,5})
SFR=(89.2.0.{1,2})
OVH=(91.121.16{1.184,4.227})

let i=0
while [[ $i -lt 200 ]] # Adapter la valeur au besoin.
do
    for RESOLVER in ${Bouygues[@]} # à remplacer par SFR ou OVH au besoin.
    do
        RAND=$(echo $[ $RANDOM % 9999999 ])
        VAR=$(echo $RAND$RAND$RAND | cut -b 1-7) # pas très élégant, je sais
        host 1."${VAR}".bash.ws "${RESOLVER}" >/dev/null 2>&1
        curl -m 5 -s https://bash.ws/dnsleak/test/"${VAR}"?json 2>/dev/null | jq -r '.[1].ip' 2>/dev/null 
    done 
    let i++
done | sort -uV | tee -a "${LOGFILE}"
```
Après, il faut éventuellement faire un peu de tri dans les résultats au cas où on soit tombé par hasard sur un numéro de test déjà fait depuis un autre FAI, bref vérifier les IP dans whois.

**Comment by Jourdelune** [comment]
Pour mon opérateur (bouyge), voilà le résultat :
```
./test.sh
62.34.202.192
62.34.202.193
62.34.202.200
62.34.202.201
62.34.202.204
62.34.202.205
62.34.202.208
62.34.202.209
62.34.202.212
62.34.202.213
62.34.202.216
62.34.202.217
87.88.192.136
87.88.192.137
87.88.192.138
87.88.192.139
87.88.192.140
87.88.192.141
87.88.192.142
87.88.192.143
2001:860:de01:3::e2
2001:860:de01:3::e7
2001:860:de09:3000::e1
2001:860:de09:3000::e8
```

**Comment by matrixbx** [comment]
Qui est une ip Orange (abonné en se fiant au reverse dns)

**Comment by ostix360** [comment]
Un fix a été mit sur la dev branch qui devrait passer sur la main branch vous pouvez tester si le site vous indique encore vulnérable ou non mais normalement il y a une verifications 
Après ceux qui ajoute des DNS local le savent général donc est ce vraiment important de vouloir vérifier cette possibilité?
